app/controllers/businessdetails/CorporationTaxRegisteredController.scala

```diff
@@ -47,33 +48,25 @@ class CorporationTaxRegisteredController @Inject () (
   val failedResult = InternalServerError("Failed to update the business corporation tax number")
 
   def get(edit: Boolean = false) = Authorised.async {
-    implicit authContext => implicit request =>
-      filterByBusinessType { cache =>
-        cache.getEntry[BusinessDetails](BusinessDetails.key) match {
-          case _ =>
-            (for {
-              bm <- OptionT.fromOption[Future](cache.getEntry[BusinessMatching](BusinessMatching.key))
-              details <- OptionT.fromOption[Future](bm.reviewDetails)
-              utr <- OptionT.fromOption[Future](details.utr)
-              _ <- updateCache(cache, CorporationTaxRegisteredYes(utr))
-            } yield getRedirectLocation(edit)) getOrElse Ok(corporation_tax_registered(EmptyForm, edit))
+    implicit authContext =>
+      implicit request =>
+        filterByBusinessType { cache =>
+          cache.getEntry[BusinessDetails](BusinessDetails.key) match {
+            case _ =>
+              (for {
+                bm <- OptionT.fromOption[Future](cache.getEntry[BusinessMatching](BusinessMatching.key))
+                details <- OptionT.fromOption[Future](bm.reviewDetails)
+                // Only update Business Details from Business Matching where it exists; after that its maintained by API5
+                _ <- if (details.utr.isDefined) {
+                  updateCache(cache, CorporationTaxRegisteredYes(details.utr.getOrElse(
+                    throw new Exception("[CorporationTaxRegisteredController][get]: Could not retrieve UTR from Business Matching")
+                  )))
+                } else {
+                  OptionT.fromOption[Future](Some(cache))
+                }
+              } yield getRedirectLocation(edit)) getOrElse InternalServerError("Could not route from CorporationTaxRegisteredController")
```

Comment:
User should never be able to edit this page. Could use Redirect(routes.ConfirmRegisteredOfficeController.get()) directly here and remove the getOrElse?

Reply:
Totally take your point on the edit however; the get or else is an action based on cache retrieval and redirect and not just redirect so I think it will need to stay as belt and braces. It handles the eventuality that something might not complete successfully in the cache related futures. Should never be hit but worth logging if it is.